### ISSUE TITLE
Prediction result : Tranformer4rec Test prediction without triton library

### ISSUE THREAD

**Issue by DilipKumar3** [NVIDIA-Merlin/Transformers4Rec#605]
# ❓ Questions & Help
I have successfully build the transformer4rec model using train data. now i want know how the recommendations look . need to see top 5,10 recommendations for each user and each instance. comparing this with my actual data. 

and need to see the affinity score for each recommendation.

Is there a way to check this right after the model building.

**Comment by rnyak** [comment]
You can do local (offline) inference meaning, you can generate topK recommendations for your test set via:
```
test_paths=<set the path to your test set>
trainer.args.predict_top_k = 20  # you can change this value
prediction = trainer.predict(test_paths)
print(prediction)
print(prediction.predictions)

```
if you want to generate ndcg and recall metrics for your topK results, you need to write your own code based on the results coming from the code above to calculate these metrics.  you can see how the metrics are calculated [here](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/transformers4rec/torch/ranking_metric.py).

**Comment by DilipKumar3** [comment]
Thanks.. this would do

**Comment by DilipKumar3** [comment]
i ran the code for prediction task but ended with error while initiating 
ENVIRONMENT : Kaggle Notebook
This my test data stored as parquet file 
![image](https://user-images.githubusercontent.com/48671630/215419441-88a48722-f1c3-49fc-9229-02e53aa79f68.png)

![image](https://user-images.githubusercontent.com/48671630/215420095-663a693d-4116-457e-8ffd-9c49055f2294.png)

This is the error im getting while running the prediction code

![image](https://user-images.githubusercontent.com/48671630/215419623-06a6a49b-afdb-490c-9bad-9376b098b9ef.png)

**Comment by rnyak** [comment]
@DilipKumar3 you need to transform your data with NVTabular first. I see your test dataset is not transformed.

**Comment by rohan-rajput-01** [comment]
Hi @DilipKumar3 , 
I was having the same issue. Try this instead 
test_paths=<test parquet file location>
trainer.test_dataset_or_path = test_paths
trainer.args.predict_top_k = 20
prediction = trainer.predict(test_paths)

cc: @rnyak

**Comment by DilipKumar3** [comment]
Thanks @rohan-rajput-01  . Its working 
How can i interpret the predictions. its has arrays of values. @rnyak

**Comment by rohan-rajput-01** [comment]
@DilipKumar3 try this 
```
print(prediction)
print(prediction.predictions)
```

**Comment by DilipKumar3** [comment]
Thanks @rohan-rajput-01  This was very helpful . 
just to confirm my understanding , prediction gives three arrays **labels**, **predictions.prediction** contains two arrays.
**lables** = session id
**predictions.predictions [0]** = recommendated items (top k ) sorted in the order.
**prediction.predictions[1]** = score for each predictions (lower the values better the model) 
@rnyak

**Comment by rnyak** [comment]
@DilipKumar3  for your comment `prediction.predictions[1] = score for each predictions (lower the values better the model)` the higher the results meaning this product has higher chance (score) to be recommended.  So you would want to recommend top K items that has the highest scores.

**Comment by DilipKumar3** [comment]
@rnyak  Thanks for the clarification. 

![image](https://user-images.githubusercontent.com/48671630/215963982-3f685007-ba7d-42d1-b8b0-c2e4dfc2c022.png)

Above table is the one record of **prediction.predictions**, 
**658** is top 1 recommended item for the user label **8456**. But all the scores are negative. 
is there specific reason for that ?

**Comment by DilipKumar3** [comment]
@rnyak i have asked a clarification of one other question that i asked above in [https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/607]()
please give reply for that

**Comment by rnyak** [comment]
> @rnyak i have asked a clarification of one other question that i asked above in https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/607 please give reply for that

Yes, I responded there already :)

**Comment by alan-ai-learner** [comment]
> @rnyak Thanks for the clarification.
> 
> ![image](https://user-images.githubusercontent.com/48671630/215963982-3f685007-ba7d-42d1-b8b0-c2e4dfc2c022.png)
> 
> Above table is the one record of **prediction.predictions**, **658** is top 1 recommended item for the user label **8456**. But all the scores are negative. is there specific reason for that ?

@rnyak Can you please explain this.
thanks!!

**Comment by rnyak** [comment]
@alan-ai-learner  are these scores coming from your own dataset or from our example notebooks? 

What you getting are the logit values, they are not probabilities. so they can be negatives. Furthermore, we use `torch.nn.LogSoftMax` (see code [here](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/transformers4rec/torch/model/prediction_task.py#L446-L463))  that the returns tensors values in the range [-inf, 0).

hope that helps.

**Comment by alan-ai-learner** [comment]
On your dataset, I'm getting this.

Thanks for the explanation.

**Comment by rnyak** [comment]
Closing this ticket since there is not recent activity. You can reopen it if necessary.

**Comment by ravkl** [comment]
It's not working for me @rnyak 
<img width="1398" alt="Screenshot 2023-06-15 at 15 46 12" src="https://github.com/NVIDIA-Merlin/Transformers4Rec/assets/46690265/bdcac2b7-515f-4f28-850f-0ee8e09db154">
